rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+
+**'`link`'** A link's payload might be a url string.
+
+**'`diagnostic`'** A diagnostic's payload might be an object including
+a level (e.g. warning, error, note), a short message string, and
+a longer detail string.
+
+
+#### The frontend chooses how to display annotations based on their type.
+
+The type of an annotation (and its documentation) should hint at how it
+might be displayed, but this choice is ultimately up to the frontend.
+Take for example a diagnostic warning annotation; a gui frontend might
+choose to show the warning inline, drawing over the text, but a tui
+frontend might only show an indicator in the gutter area:
+
+![vscode language server error](./assets/vscode_err.png)<br> _A language
+server error in Visual Studio Code_ <br>
+
+![neovim language server error](./assets/nvim_err.png)<br> _A language
+server error in Neovim_
+
+![Xcode error](./assets/xcode_err.png)<br> _An error in Xcode, just for
+variety_
+
+### plugin API
+
+Plugins add, update, and remove annotations with a single protocol
+method:
+
+```javascript
+{
+    "method": "update_annotations",
+    "params": {
+        "type": "diagnostics",
+        "range": [0, 1337],
+        "items": [ /* some annotations */],
+    }
+}
+```
+
+Each plugin can modify its own annotations. Each annotation _type_ is
+stored separately.
+
+#### Plugins update annotations during editing.
+
+Annotations will be adjusted or removed in core, during editing. (The
+particulars of this are undecided; there may be different update
+strategies for different annotation types, or we may just destroy
+annotations when an edit occurs in their interior.) In either case, the
+plugin is expected to remove or update any existing annotations in
+response to a user's edits.
+
+### Example: diagnostic messages
+
+Consider the case of a a linting / diagnostic plugin; in particular
+let's imagine a simple plugin that wraps the  `cargo check` command,
+running it in the background when a file is saved.
+
+In the case of our `cargo check` plugin, it would run `cargo check
+--message-format=json`, capture the output, and then extract it into
+something like:
+
+```javascript
+// AnnotationSet
+{
+    “type”: "diagnostic",
+    "ranges": [[420, 456]],
+    “payloads”: [
+        {
+            "level": "warning", // warning / error / info / ?
+            "message": "unused import: `std::marker::PhantomData`",
+            "code": "unused_import"
+    ],
+    “metadata”: null
+}
+```
+This annotation set would be sent to core, which would then forward it
+to the frontend to display as desired.
+
+## Questions / considerations
+
+A number of things have come up while drafting this proposal that are
+worth noting.
+
+#### metadata handling
+
+We probably want (and have assumed having,elsewhere in this proposal)
+some way to associate metadata with a _set_ of annotations;
+for instance 'find' annotations might include the total
+number of results and their approximate locations. This does add some
```

Comment:
Right, here's the discussion of total count. My personal sense is that it makes sense for it to be part of the generic update protocol and not type-specific metadata, but I can be talked out of it.

Reply:
Even if we include 'total count' as a general parameter, we would still likely want generic metadata for other purposes; for instance find may want to indicate which of the `n` total results is currently selected. I don't feel like I have a particularly clear answer here, right now; the metadata question will probably need to get hashed out further as we explore concrete use cases.